crates/sync/alchemy/src/client.rs

```diff
@@ -10,32 +10,45 @@ use serde_json::json;
 use crate::{networks, types::AlchemyAssetTransfer, Result};
 
 #[derive(Debug)]
-pub(crate) struct Client(Provider<RetryClient<Http>>);
+pub(crate) struct Client {
+    provider: Provider<RetryClient<Http>>,
+    chain_id: u32,
+    api_key: String,
+}
 
 pub(crate) enum Direction {
     From(Address),
     To(Address),
 }
 
 impl Client {
-    pub fn new(chain_id: u32, api_key: &str) -> Result<Self> {
-        let endpoint = networks::get_endpoint(chain_id, api_key)?;
-        let http = Http::new(endpoint);
-
-        let policy = Box::<HttpRateLimitRetryPolicy>::default();
-
-        let res = RetryClientBuilder::default()
-            .rate_limit_retries(10)
-            .timeout_retries(3)
-            .initial_backoff(Duration::from_millis(500))
-            .compute_units_per_second(300)
-            .build(http, policy);
+    pub fn new(chain_id: u32, api_key: &str, path: &str) -> Result<Self> {
+        let provider = {
+            let endpoint = networks::get_endpoint(chain_id, path, api_key)?;
+            let http = Http::new(endpoint);
+            let policy = Box::<HttpRateLimitRetryPolicy>::default();
+
+            let res = RetryClientBuilder::default()
+                .rate_limit_retries(10)
+                .timeout_retries(3)
+                .initial_backoff(Duration::from_millis(500))
+                .compute_units_per_second(300)
+                .build(http, policy);
+
+            Provider::new(res)
+        };
 
-        Ok(Self(Provider::new(res)))
+        Ok(Self {
+            provider,
+            chain_id,
+            api_key: api_key.to_string(),
+        })
     }
 
     pub async fn get_block_number(&self) -> Result<U64> {
-        let block = self.0.get_block_number().await?;
+        let client = Client::new(self.chain_id, &self.api_key, "v2/");
```

Comment:
something's off here. You already have `self` which is a `client`, and you're now instantiating a new `Client` only to give it a different path

I understand you need to, in some circustances swap that last part of the URL, but this isn't the way to do it

you can have instantiate two providers on `new`, one regular for the previous behaviour, and an `nft_v3_provider` for the separate endpoint, and then use it where needed